lib/puppet/functions/reduce.rb

```diff
@@ -1,61 +1,94 @@
-#       $a.reduce |$memo, $entry| { [sum, $memo[1]+$entry[1]] }
-#       #=> [sum, 6]
-#
-#       # reverse a string
-#       "abc".reduce |$memo, $char| { "$char$memo" }
-#       #=>"cbe"
-#
-# It is possible to provide a starting 'memo' as an argument.
-#
-# @example Using reduce with given start 'memo'
-#
-#       # Reduce an array
-#       $a = [1,2,3]
-#       $a.reduce(4) |$memo, $entry| { $memo + $entry }
-#       #=> 10
-#
-#       # Reduce hash values
-#       $a = {a => 1, b => 2, c => 3}
-#       $a.reduce([na, 4]) |$memo, $entry| { [sum, $memo[1]+$entry[1]] }
-#       #=> [sum, 10]
-#
-# @example Using reduce with an Integer range
-#
-#       Integer[1,4].reduce |$memo, $x| { $memo + $x }
-#       #=> 10
+# Applies a [lambda](https://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html)
+# to every value in a data structure from the first argument, carrying over the returned
+# value of each iteration, and returns the result of the lambda's final iteration. This
+# lets you create a new value or data structure by combining values from the first
+# argument's data structure.
+#
+# This function takes two mandatory arguments, in this order:
+# 1. An array or hash the function will iterate over.
+# 2. A lambda, which the functions calls for each element in the first argument.
+#
+# The lambda takes two parameters, in this order:
+# 1. A memo value that is overwritten after each iteration with the iteration's result.
+# 2. A second value that is overwritten after each iteration with the next value in the
+# function's first argument.
+#
+# @example Using the `reduce` function
+#
+# `$data.reduce |$memo, $value| { ... }`
+#
+# or
+#
+# `reduce($data) |$memo, $value| { ... }`
+#
+# You can also pass an optional "start memo" value as an argument, such as `start` below:
+#
+# `$data.reduce(start) |$memo, $value| { ... }`
+#
+# or
+#
+# `reduce($data, start) |$memo, $value| { ... }`
+#
+# When the first argument (`$data` in the above example) is an array, Puppet passes each
+# of the data structure's values in turn to the lambda's parameters. When the first
+# argument is a hash, Puppet converts each of the hash's values to an array in the form
+# `[key, value]`.
+#
+# If you don't pass a start memo, Puppet passes the data structure's first two values to
+# the lambda in order, then executes the lambda. If you pass a start memo, Puppet instead
+# passes it and the data structure's first value to the lambda's parameters before its
+# first iteration.
+#
+# The lambda returns each iteration's result to the memo parameter (`$memo` in the above
+# examples), then passes the data structure's next value to the lambda's second parameter
+# (`$value` above). Puppet repeats the process for each of the data structure's remaining
+# values.
+#
+# If the structure has one value, Puppet returns the value and does not process the
+# lambda.
+#
+# @example Using the `reduce` function with a two-parameter lambda
```

Comment:
Since you can't use a one- or three-parameter lambda, I'd drop that part and distinguish the examples by just the presence of the start memo.